clang/lib/AST/DeclCXX.cpp

```diff
@@ -857,8 +881,13 @@ void CXXRecordDecl::addedMember(Decl *D) {
         data().HasDeclaredCopyAssignmentWithConstParam = true;
     }
 
-    if (Method->isMoveAssignmentOperator())
+    if (Method->isMoveAssignmentOperator()) {
       SMKind |= SMF_MoveAssignment;
+    }
+
+    if (Method->isUserProvided() &&
+        (Method->isCopyAssignment() || Method->isMoveAssignment()))
+      data().IsNaturallyTriviallyRelocatable = false;
```

Comment:
If we stick to the 2786r4 definition, this is overly restrictive (ie assignment does not affect relocatability)

Reply:
Some solicited comments, from Qt maintainer.

I don't have a strong opinion on whether the presence of an assignment operator should affect relocatability. Only "stupid types" will have a trivial copy/move constructor and a non-trivial assignment operator that does something different. Those types are inherently broken, because a move can be implemented as either
```c++
   tgt = std::move(src);
```
or as
```c++
    std::destroy_at(&tgt);
    std::construct_at(&tgt, std::move(src));
```
(provided `&tgt != &src`)

It is the container implementer's choice of which those two to use and we have both being chosen in different frameworks, sometimes even both inside the same framework. Therefore, any type that has visibly different behaviour for the two implementations above is "stupid" and has bigger problems than its relocatability. 

That means it's a reasonable choice to say the presence of the assignment operator does not affect the relocatability. But then it's also an equally valid choice to say it does, with the benefit that one could find out that a type is "stupid" before a release.

I do have a weak "leans to" choice though: for the first implementation of trivial relocatability, we should err on the side of being more restrictive. We can always be less so in the future, but going the other direction is more problematic.

Reply:
I've also been asked to comment. My personal opinion is that relocability is strictly tied to the Rule Of Five. The moment a user decides to opt into the RO5 by e.g. writing a custom assignment operator, they should lose trivial relocability, unless they do something extra to gain it back (different proposals disagree on the syntax but ultimately they achieve the same). (We may then argue regarding all the subtleties about user-declared vs. user-defaulted vs. user-defined, but I don't really have a strong opinion about that.)

Reply:
My concern here is to make sure we are consistent with WG21.
EWG approved a design that does not consider assignment, and Clang should follow suite (ie we should avoid needing 2 type traits, one that is standard conforming, and one that isn't).

Either way we have some time to adapt that, as this paper should (finger crossed) be approved in june, before we need to ship clang 18.

I generally agree with @thiagomacieira (Hi!) that this is mostly a question for stupid types, but I think the good mental model is that relocation is an optimization of copy. IE a type that does not follow the the rule of 5 is "stupid" whether it's relocatable or not.

So my advice here is do as P2786 does, and revisit should that paper fail to be adopted before the next clang release.

Reply:
> Only "stupid types" will have a trivial copy/move constructor and a non-trivial assignment operator that does something different.

One such type is `std::tuple<int&>`. As far as I can tell, the way the different papers handle this type is:

* P1144: not trivially relocatable (because the copy/move assignment operator is user-provided)
* P2786: trivially relocatable (because we only care about move construction, and that one is defaulted)

I don't actually know what goes wrong (if anything) if you make `tuple<int&>` trivially relocatable, but at least that's a nice concrete example to think about.